### PR TITLE
feat: palm integration

### LINKED ISSUE #27: feat: Palm integration
# Description

We aim to enhance the capabilities of our AI Gateway by integrating Google's PaLM LLM Model, commonly known as "Palm," into the existing infrastructure. This issue serves as a discussion and tracking point for the integration efforts.

### Integration Tasks:

- **Model Integration**: Develop a component to integrate Google's Palm Detection Model into the AI Gateway. This component will handle the request reponse of the inference APIs from the model.

- **Pre-processing and Post-processing**: Implement necessary pre-processing and post-processing steps.

- **API and Input/Output Design**: Define an API for interacting with the Palm Detection Model within the AI Gateway, specifying input parameters and expected output format.

- **Documentation**: Update the AI Gateway's documentation to include information on how to use the newly integrated Palm LLM Model.

### PR DESCRIPTION
**Pull request by noble-varghese**
Fixes: #27

### CODE REVIEW

File: src/providers/palm/complete.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        min: 0,
+        max: 1,
+    },
+    top_k: {
+        param: "topK",
+        default: 1,
+        min: 0,
+        max: 1,
+    },
+    candidate_count: {
```
Comment (VisargD) [inline src/providers/palm/complete.ts]: candidate_count seems like an equivalent to `n` which is there in existing specification. Please reuse that and make the transformation to palm specific key

File: src/providers/palm/chatComplete.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        min: 0,
+        max: 1,
+    },
+    top_k: {
+        param: "topK",
+        default: 1,
+        min: 0,
+        max: 1,
+    },
+    candidate_count: {
```
Comment (VisargD) [inline src/providers/palm/chatComplete.ts]: candidate_count seems like an equivalent to n which is there in existing specification. Please reuse that and make the transformation to palm specific key

File: src/providers/palm/chatComplete.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import { PalmChatCompleteResponse } from "../../types/responseBody";
+import { ChatCompletionResponse, ErrorResponse, ProviderConfig } from "../types";
+
+// TODOS: this configuration does not enforce the maximum token limit for the input parameter. If you want to enforce this, you might need to add a custom validation function or a max property to the ParameterConfig interface, and then use it in the input configuration. However, this might be complex because the token count is not a simple length check, but depends on the specific tokenization method used by the model.
+
+export const PalmChatCompleteConfig: ProviderConfig = {
+    model: {
+        param: "model",
+        required: true,
+        default: "gpt-3.5-turbo",
```
Comment (VisargD) [inline src/providers/palm/chatComplete.ts]: Please fix the default model name

File: src/providers/palm/complete.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import { PalmCompleteResponse } from "../../types/responseBody";
+import { ChatCompletionResponse, ErrorResponse, ProviderConfig } from "../types";
+
+// TODOS: this configuration does not enforce the maximum token limit for the input parameter. If you want to enforce this, you might need to add a custom validation function or a max property to the ParameterConfig interface, and then use it in the input configuration. However, this might be complex because the token count is not a simple length check, but depends on the specific tokenization method used by the model.
+
+export const PalmCompleteConfig: ProviderConfig = {
+    model: {
+        param: "model",
+        required: true,
+        default: "gpt-3.5-turbo",
```
Comment (VisargD) [inline src/providers/palm/complete.ts]: Please fix the default model name

File: src/providers/palm/complete.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    candidate_count: {
+        param: "candidateCount",
+        default: 1,
+        min: 1,
+        max: 8
+    }
+};
+
+export const PalmCompleteResponseTransform: (response: PalmCompleteResponse, responseStatus: number) => ChatCompletionResponse | ErrorResponse = (response, responseStatus) => {
+    console.log("responseStatus: ", responseStatus)
```
Comment (VisargD) [inline src/providers/palm/complete.ts]: Please remove console log

File: src/providers/palm/complete.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    if (responseStatus !== 200) {
+        return {
+            error: {
+                message: response.error?.message ?? null,
+                type: null,
+                param: null,
+                code: response.error?.code ?? null
+            },
+            provider: "anthropic"
```
Comment (VisargD) [inline src/providers/palm/complete.ts]: Please fix the provider name

File: src/providers/palm/embed.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+            if (Array.isArray(params.input)) {
+                return params.input;
+            } else {
+                return [params.input];
+            }
+        }
+    },
+    model: {
+        param: "model",
+        default: "embed-english-light-v2.0",
```
Comment (VisargD) [inline src/providers/palm/embed.ts]: This is not a palm model. Please fix

File: src/providers/palm/index.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import { PalmEmbedConfig, PalmEmbedResponseTransform } from "./embed";
+
+const PalmAIConfig: ProviderConfigs = {
+  complete: PalmCompleteConfig,
+  embed: PalmEmbedConfig,
+  api: PalmApiConfig,
+  chatComplete: PalmChatCompleteConfig,
+  responseTransforms: {
+    complete: PalmCompleteResponseTransform,
+    // 'stream-complete': PalmCompleteResponseTransform,
```
Comment (VisargD) [inline src/providers/palm/index.ts]: Please remove comments if stream mode is not supported

File: src/providers/palm/index.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    chatComplete: PalmChatCompleteResponseTransform,
+    // 'stream-chatComplete': PalmChatCompleteResponseTransform,
```
Comment (VisargD) [inline src/providers/palm/index.ts]: Please remove comments if stream mode is not supported

File: src/providers/palm/embed.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+export const PalmEmbedResponseTransform: (response: PalmEmbedResponse) => EmbedResponse = (response) => {
+    return {
+    object: "list",
+    data: [{
+        object: "embedding",
+        embedding: response.embedding.value,
+        index: 0,
+    }],
+    model: "", // Todo: find a way to send the cohere embedding model name back
```
Comment (VisargD) [inline src/providers/palm/embed.ts]: Please fix/remove the comment text

**Comment by VisargD** [comment]
@noble-varghese - Please also add the support in proxy route. And please also do the same for tryPostProxy function to support it in config + proxy calls

File: src/providers/palm/api.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import { ProviderAPIConfig } from "../types";
+
+export const PalmApiConfig: ProviderAPIConfig = {
+    baseURL: "https://generativelanguage.googleapis.com/v1beta3",
+    headers: () => {
+        return { "Content-Type": "application/json" }
+    },
+    getEndpoint: (fn: string, API_KEY: string, model: string) => {
```
Comment (VisargD) [inline src/providers/palm/api.ts:8]: API key is not required here to determine endpoint. Please remove it and also modify wherever this is called.

File: src/providers/palm/chatComplete.ts [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        min: 0,
+        max: 1,
+    },
+    top_k: {
+        param: "topK",
+        default: 1,
+        min: 0,
+        max: 1,
+    },
+    n: {
```
Comment (VisargD) [inline src/providers/palm/chatComplete.ts:48]: candidate count is in snake case as per PALM docs.
https://developers.generativeai.google/tutorials/curl_quickstart

File: src/providers/palm/complete.ts [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        max: 1,
+    },
+    top_k: {
+        param: "topK",
+        default: 1,
+        min: 0,
+        max: 1,
+    },
+    n: {
+        param: "candidateCount",
```
Comment (VisargD) [inline src/providers/palm/complete.ts:43]: candidate count is in snake case as per PALM docs.
https://developers.generativeai.google/tutorials/curl_quickstart

File: src/providers/palm/chatComplete.ts [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        param: "model",
+        required: true,
+        default: "model/chat-bison-001",
+    },
+    messages: {
+        param: "prompt",
+        default: "",
+        transform: (params: Params) => {
+            const { examples, context, messages } = params
+            const palm_messages = messages?.map(message => ({
```
Comment (VisargD) [inline src/providers/palm/chatComplete.ts]: Please use camelCase for codebase

File: src/providers/palm/chatComplete.ts (hunk 12 above)
Comment (noble-varghese) [reply src/providers/palm/chatComplete.ts:48]: Candidate count is in CamelCase according to this doc - https://developers.generativeai.google/api/rest/generativelanguage/models/generateText#request-body
This is the detailed API usage docs.

File: src/providers/palm/api.ts (hunk 11 above)
Comment (noble-varghese) [reply src/providers/palm/api.ts:8]: API key is a query param here and is part of the URL

File: src/providers/palm/chatComplete.ts (hunk 12 above)
Comment (VisargD) [reply src/providers/palm/chatComplete.ts:48]: Okay. Looks like an issue on their docs. We can close this if camelCase is working fine without any errors

File: src/providers/palm/api.ts (hunk 11 above)
Comment (VisargD) [reply src/providers/palm/api.ts:8]: Ohh. Got it.

File: src/providers/palm/complete.ts (hunk 13 above)
Comment (VisargD) [reply src/providers/palm/complete.ts:43]: Closing as per this https://github.com/Portkey-AI/Rubeus/pull/30/files#r1374132451

**Comment by noble-varghese** [comment]
> @noble-varghese - Please also add the support in proxy route. And please also do the same for tryPostProxy function to support it in config + proxy calls

Support for the proxy is there by default. Let me know if there is anything that might be required additionally to support this.

**Review by VisargD — APPROVED** [review]

**Pull request merged**